File: Painter/canvas.c
//
//  canvas.c
//  Painter
//
//  Created by sodas on 10/31/16.
//  Copyright © 2016 sodastsai. All rights reserved.
//

#include "canvas.h"
#include <stdlib.h>

CVCanvasPoint const kCVCanvasPointOrigin = {.x = 0, .y = 0};

#pragma mark - Rect

CVCanvasRect CVCanvasRectMake(double x, double y, double width, double height) {
    return (CVCanvasRect){.origin = (CVCanvasPoint){.x = x, .y = y},
                          .size = (CVCanvasSize){.width = width, .height = height}};
}

CVCanvasPoint CVCanvasRectGetCenter(CVCanvasRect rect) {
    return (CVCanvasPoint){.x = rect.origin.x + rect.size.width/2,
                           .y = rect.origin.y + rect.size.height/2};
}

void CVCanvasRectSetCenter(CVCanvasRect rect, CVCanvasPoint center) {
    rect.origin = (CVCanvasPoint){.x = center.x - rect.size.width/2, .y = center.y - rect.size.height/2};
}

#pragma mark - Color

CVCanvasColor const kCVCanvasColorWhite  = {.r=1.0, .g=1.0, .b=1.0, .a=1.0};
CVCanvasColor const kCVCanvasColorBlack  = {.r=0.0, .g=0.0, .b=0.0, .a=1.0};
CVCanvasColor const kCVCanvasColorRed    = {.r=1.0, .g=0.0, .b=0.0, .a=1.0};
CVCanvasColor const kCVCanvasColorOrange = {.r=1.0, .g=0.5, .b=0.0, .a=1.0};
CVCanvasColor const kCVCanvasColorYellow = {.r=1.0, .g=1.0, .b=0.0, .a=1.0};
CVCanvasColor const kCVCanvasColorGreen  = {.r=0.0, .g=1.0, .b=0.0, .a=1.0};
CVCanvasColor const kCVCanvasColorCyan   = {.r=0.0, .g=1.0, .b=1.0, .a=1.0};
CVCanvasColor const kCVCanvasColorBlue   = {.r=0.0, .g=0.0, .b=1.0, .a=1.0};
CVCanvasColor const kCVCanvasColorPurple = {.r=1.0, .g=0.0, .b=1.0, .a=1.0};

CVCanvasColor CVCanvasColorFromRGBAValues(const double values[]) {
    return (CVCanvasColor){.r = values[0], .g = values[1], .b = values[2], .a = values[3]};
}

#pragma mark - Drawable

CVCanvasDrawable CVCanvasDrawableFromRect(CVCanvasRect rect) {
    return (CVCanvasDrawable){.rect = rect,
                              .fillColor = kCVCanvasColorWhite,
                              .strokeColor = kCVCanvasColorBlack,
                              .strokeWidth = 1.0 };
}

CVCanvasDrawable CVCanvasDrawableRevertColor(CVCanvasDrawable drawable) {
    return (CVCanvasDrawable){.rect = drawable.rect,
                              .fillColor = drawable.strokeColor,
                              .strokeColor = drawable.fillColor,
                              .strokeWidth = drawable.strokeWidth };
}

static inline double CVColorRandomChannelValue() {
    uint32_t randomValue = arc4random_uniform(255);
    return randomValue/255.0;
}

static inline CVCanvasColor CVColorRandomColor() {
    return (CVCanvasColor){.r = CVColorRandomChannelValue(),
                           .g = CVColorRandomChannelValue(),
                           .b = CVColorRandomChannelValue(),
                           .a = CVColorRandomChannelValue()/2+0.5};
}

CVCanvasDrawable CVCanvasDrawableCreateRandomly(double xBound, double yBound, double widthBound, double heightBound) {
    CVCanvasColor fillColor = CVColorRandomColor();
    CVCanvasColor strokeColor = CVColorRandomColor();
    double strokeWidth = arc4random_uniform(20) + 1;
    CVCanvasPoint origin = {.x = arc4random_uniform(xBound), .y = arc4random_uniform(yBound)};
    CVCanvasSize size = {.width = arc4random_uniform(widthBound), .height = arc4random_uniform(heightBound)};
    return (CVCanvasDrawable){.rect = (CVCanvasRect){.origin = origin, .size = size},
                              .fillColor = strokeColor,
                              .strokeColor = fillColor,
                              .strokeWidth = strokeWidth};
}

File: Painter/canvas.swift
//
//  canvas.swift
//  Painter
//
//  Created by sodas on 10/31/16.
//  Copyright © 2016 sodastsai. All rights reserved.
//

import CoreGraphics.CGContext
import AppKit.NSColor

// MARK: - Refined for Swift

extension CanvasColor {
    init?(from values: [Double]) {
        guard values.count == 4 else {
            return nil
        }
        var _values = values
        self = __CVCanvasColorFromRGBAValues(&_values)
    }
}

// MARK: - Swift Protocols and Helpers

extension CanvasPoint: CustomStringConvertible {
    public var description: String {
        return "(\(self.x), \(self.y))"
    }
}
extension CGPoint {
    init(_ canvasPoint: CanvasPoint) {
        self.init(x: canvasPoint.x, y: canvasPoint.y)
    }
}

extension CanvasSize: CustomStringConvertible {
    public var description: String {
        return "(w=\(self.width), h=\(self.height))"
    }
}
extension CGSize {
    init(_ canvasSize: CanvasSize) {
        self.init(width: canvasSize.width, height: canvasSize.height)
    }
}

extension CanvasRect: CustomStringConvertible {
    public var description: String {
        return "(origin=\(self.origin), size=\(self.size))"
    }
}
extension CGRect {
    init(_ canvasRect: CanvasRect) {
        self.init(origin: CGPoint(canvasRect.origin), size: CGSize(canvasRect.size))
    }
}

extension CanvasColor: CustomStringConvertible {
    public var description: String {
        return "{r=\(self.r), g=\(self.g), b=\(self.b), a=\(self.a)}"
    }
}
extension CanvasColor {
    var nsColor: NSColor {
        return NSColor(red: CGFloat(self.r),
                       green: CGFloat(self.g),
                       blue: CGFloat(self.b),
                       alpha: CGFloat(self.a))
    }
    var cgColor: CGColor {
        return self.nsColor.cgColor
    }
}

// MARK: - Canvas Draw

protocol Drawable {
    func draw(with context: CGContext)
}

extension CanvasDrawable: Drawable {
    func draw(with context: CGContext) {
        context.saveGState()
        defer {
            context.restoreGState()
        }

        context.setFillColor(self.fillColor.cgColor)
        context.fill(CGRect(self.rect))
        context.setStrokeColor(self.strokeColor.cgColor)
        context.stroke(CGRect(self.rect), width: CGFloat(self.strokeWidth))
    }
}

extension Collection where Self.Iterator.Element: Drawable {
    func draw(with context: CGContext) {
        for drawable in self {
            context.saveGState()
            defer {
                context.restoreGState()
            }
            drawable.draw(with: context)
        }
    }
}
